### PR TITLE
OHRM5X-287: Develop OpenID - provider screens

### PR DESCRIPTION
**Pull request by ManulMax**

### CODE REVIEW

File: build/build.xml [hunk 1]
```diff
@@ -70,7 +70,6 @@
 
         <exclude name="src/plugins/orangehrmAuthenticationPlugin/lib/**"/>
         <exclude name="src/plugins/orangehrmCoreOAuthPlugin/lib/**"/>
```
[the remark this answers is not in the document]
Comment (ManulMax) [reply build/build.xml:72]: there is no folder realted to that path

File: installer/Migration/V5_4_0/Migration.php [hunk 2]
```diff
@@ -36,7 +37,7 @@ public function up(): void
     {
         $groups = ['admin', 'auth', 'general'];
         foreach ($groups as $group) {
-            $this->getLangStringHelper()->insertOrUpdateLangStrings($group);
+            $this->getLangStringHelper()->insertOrUpdateLangStrings(__DIR__, $group);
```
[the remark this answers is not in the document]
Comment (ManulMax) [reply installer/Migration/V5_4_0/Migration.php:40]: missed one

File: src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/AddProvider.vue [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        ],
+        clientId: [required, shouldNotExceedCharLength(255)],
+        clientSecret: [required, shouldNotExceedCharLength(255)],
+        url: [required, validURL],
+      },
+    };
+  },
+  methods: {
+    onCancel() {
+      navigate('/admin/openIdProvider');
```
[the remark this answers is not in the document]
Comment (ManulMax) [reply src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/AddProvider.vue:135]: same as 4X

File: src/client/src/core/util/validation/rules.ts [hunk 4]
```diff
@@ -700,3 +700,11 @@ export const digitsOnlyWithTwoDecimalPoints = function (
     translate('claim.should_be_a_valid_number')
   );
 };
+
+export const validURL = function (value: string): boolean | string {
+  return (
+    !value ||
+    /^(https?|ftp):\/\/[^\s/$.?#].[^\s]*$/.test(value) ||
```
[the remark this answers is not in the document]
Comment (ManulMax) [reply src/client/src/core/util/validation/rules.ts:707]: http allowed. will remove ftp

File: installer/Migration/V5_6_0/lang-string/admin.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+langStrings:
+  -
+    value: 'Client Secret'
+    unitId: secret
+  -
+    value: 'Developer Key'
+    unitId: developer_key
```
[the remark this answers is not in the document]
Comment (ManulMax) [reply installer/Migration/V5_6_0/lang-string/admin.yaml:7]: removed

File: installer/Migration/V5_6_0/lang-string/admin.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    value: 'Add Provider'
+    unitId: add_provider
+  -
+    value: 'Edit Provider'
+    unitId: edit_provider
+  -
+    value: 'Provider List'
+    unitId: provider_list
+  -
+    value: 'Should be a Valid URL'
```
Comment (RajithaKumara) [inline installer/Migration/V5_6_0/lang-string/admin.yaml:18]: Let's remove this validation

File: installer/Migration/V5_6_0/lang-string/admin.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
+langStrings:
+  -
+    value: 'Client Secret'
```
Comment (RajithaKumara) [inline installer/Migration/V5_6_0/lang-string/admin.yaml:3]: this is already there in OAuth
`admin.client_secret`

File: src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/AddProvider.vue [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  shouldNotExceedCharLength,
+} from '@ohrm/core/util/validation/rules';
+import useServerValidation from '@/core/util/composable/useServerValidation';
+
+const initialAuthProvider = {
+  name: '',
+  url: '',
+  clientId: '',
+  clientSecret: '',
+  status: 1,
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/AddProvider.vue:96]: hope this is optional parameter in BE API. better to set default value in BE and make it optional if not

File: src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/EditProvider.vue [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+              <oxd-input-field
+                v-model="authProvider.clientId"
+                :label="$t('admin.client_id')"
+                required
+              />
+            </oxd-grid-item>
+            <oxd-grid-item>
+              <oxd-input-field
+                v-model="authProvider.clientSecret"
+                :label="$t('admin.secret')"
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/EditProvider.vue:56]: *

File: src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/AddProvider.vue [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+                :rules="rules.clientId"
+                :label="$t('admin.client_id')"
+                required
+              />
+            </oxd-grid-item>
+            <oxd-grid-item>
+              <oxd-input-field
+                v-model="authProvider.clientSecret"
+                :rules="rules.clientSecret"
+                :label="$t('admin.secret')"
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/AddProvider.vue:59]: *

File: src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/EditProvider.vue [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+      isLoading: false,
+      secretPlaceholder: '',
+      authProvider: {
+        name: '',
+        url: '',
+        clientId: '',
+        clientSecret: '',
+      },
+      rules: {
+        name: [required, shouldNotExceedCharLength(40)],
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/EditProvider.vue:119]: seems some rules are missing

File: src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/ProviderList.vue [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+    const {
+      showPaginator,
+      currentPage,
+      total,
+      pages,
+      pageSize,
+      response,
+      isLoading,
+      execQuery,
+    } = usePaginate(httpWithStatus);
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/ProviderList.vue:96]:
```suggestion
    } = usePaginate(http, {query: {status: true}});
```
:thinking:

File: src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/ProviderList.vue [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+export default {
+  components: {
+    'delete-confirmation': DeleteConfirmationDialog,
+  },
+
+  setup() {
+    const httpWithStatus = new APIService(
+      window.appGlobal.baseUrl,
+      '/api/v2/auth/openid-providers?status=true',
+    );
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/ProviderList.vue:82]:
```suggestion
```
:thinking:

File: src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/ProviderList.vue [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+      total,
+      pages,
+      pageSize,
+      response,
+      isLoading,
+      execQuery,
+    } = usePaginate(httpWithStatus);
+    return {
+      http,
+      httpWithStatus,
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/socialMediaAuth/ProviderList.vue:99]: do we need to return these two? :thinking:

File: src/client/src/orangehrmAuthenticationPlugin/components/AuthButton.vue [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+ -->
+<script>
+import {h} from 'vue';
+import {OxdText} from '@ohrm/oxd';
+
+export default {
+  name: 'AuthButton',
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAuthenticationPlugin/components/AuthButton.vue:25]: TODO

File: src/client/src/orangehrmAuthenticationPlugin/components/SocialMediaAuth.vue [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+ -->
+<template>
+  <div>
+    <oxd-text tag="p" class="orangehrm-social-auth-header">
+      {{ $t('general.or_login_with') }}
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAuthenticationPlugin/components/SocialMediaAuth.vue:23]: why this is under `general`? is this already added?

File: src/client/src/orangehrmAuthenticationPlugin/components/SocialMediaAuth.vue [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+      socialAuthenticators,
+      http,
+    };
+  },
+
+  // TODO:: URL
+  methods: {
+    onClickAction(id) {
+      this.http.request({
+        method: 'POST',
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAuthenticationPlugin/components/SocialMediaAuth.vue:77]: TODO:: I want to check this further.

File: src/plugins/orangehrmAuthenticationPlugin/Controller/LoginController.php [hunk 18]
```diff
@@ -55,6 +61,17 @@ public function getHomePageService(): HomePageService
...
 
+    /**
+     * @return SocialMediaAuthenticationService
+     */
+    public function getSocialMediaAuthenticationService(): SocialMediaAuthenticationService
+    {
+        if (!$this->socialMediaAuthenticationService instanceof SocialMediaAuthenticationService) {
+            $this->socialMediaAuthenticationService = new SocialMediaAuthenticationService();
+        }
+        return $this->socialMediaAuthenticationService;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAuthenticationPlugin/Controller/LoginController.php:72]:
```suggestion
        return $this->socialMediaAuthenticationService ??= new SocialMediaAuthenticationService();
```
:thinking: 
OR
why we are not getting from the service container

File: src/plugins/orangehrmCorePlugin/Api/Rest/ValidationUniqueAPI.php [hunk 19]
```diff
@@ -63,7 +63,8 @@ class ValidationUniqueAPI extends Endpoint implements ResourceEndpoint
         'LeaveType' => ['attributeName' => ['name']],
         'Employee' => ['attributeName' => ['employeeId']],
         'TerminationReason' => ['attributeName' => ['name']],
-        'Vacancy' => ['attributeName' => ['name']]
+        'Vacancy' => ['attributeName' => ['name']],
```
Comment (RajithaKumara) [inline src/plugins/orangehrmCorePlugin/Api/Rest/ValidationUniqueAPI.php:66]: @devishke-orange this seems a coupling right. anyway we did it simplest way. do you have any suggestions?

File: src/plugins/orangehrmOpenidAuthenticationPlugin/Controller/OpenIdConnectLoginController.php [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+use OrangeHRM\Core\Traits\Auth\AuthUserTrait;
+use OrangeHRM\Framework\Http\Request;
+use OrangeHRM\OpenidAuthentication\Service\SocialMediaAuthenticationService;
+
+class OpenIdConnectLoginController extends AbstractVueController implements PublicControllerInterface
+{
+    use AuthUserTrait;
+
+    public const SCOPE = 'email';
+    public const REDIRECT_URL = '';
```
Comment (RajithaKumara) [inline src/plugins/orangehrmOpenidAuthenticationPlugin/Controller/OpenIdConnectLoginController.php:35]: This is not clear for me

File: src/plugins/orangehrmOpenidAuthenticationPlugin/Controller/OpenIdConnectLoginController.php [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+    /**
+     * @var SocialMediaAuthenticationService|null
+     */
+    protected ?SocialMediaAuthenticationService $socialMediaAuthenticationService = null;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmOpenidAuthenticationPlugin/Controller/OpenIdConnectLoginController.php:40]: why we are not getting from the service container

File: src/plugins/orangehrmOpenidAuthenticationPlugin/Controller/OpenIdConnectRedirectController.php [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * @return HomePageService
+     */
+    public function getHomePageService(): HomePageService
+    {
+        if (!$this->homePageService instanceof HomePageService) {
+            $this->homePageService = new HomePageService();
+        }
+        return $this->homePageService;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmOpenidAuthenticationPlugin/Controller/OpenIdConnectRedirectController.php:43]:
```suggestion
        return $this->homePageService ??= new HomePageService();
```

File: src/plugins/orangehrmOpenidAuthenticationPlugin/Controller/OpenIdConnectRedirectController.php [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    //TODO
```
Comment (RajithaKumara) [inline src/plugins/orangehrmOpenidAuthenticationPlugin/Controller/OpenIdConnectRedirectController.php:46]: TODO

File: src/plugins/orangehrmOpenidAuthenticationPlugin/Dao/AuthProviderDao.php [hunk 24]
```diff
@@ -127,4 +127,16 @@ public function saveAuthProviderExtraDetails(AuthProviderExtraDetails $authProvi
         $this->persist($authProviderExtraDetails);
         return $authProviderExtraDetails;
     }
+
+    /**
+     * @return array
```
Comment (RajithaKumara) [inline src/plugins/orangehrmOpenidAuthenticationPlugin/Dao/AuthProviderDao.php:132]:
```suggestion
     * @return OpenIdProvider[]
```

File: src/plugins/orangehrmOpenidAuthenticationPlugin/config/routes.yaml [hunk 25]
```diff
@@ -15,17 +15,23 @@ apiv2_auth_openid_provider:
 ##########################
 #        Screens
 ##########################
-add_auth_provider:
+admin_add_auth_provider:
   path: admin/addAuthProvider
   controller: OrangeHRM\OpenidAuthentication\Controller\SaveSocialMediaAuthProviderController::handle
   methods: [ GET ]
 
 edit_auth_provider:
```
Comment (RajithaKumara) [inline src/plugins/orangehrmOpenidAuthenticationPlugin/config/routes.yaml:23]:
```suggestion
admin_edit_auth_provider:
```
i assumed you will fix this as well

File: src/plugins/orangehrmOpenidAuthenticationPlugin/config/routes.yaml [hunk 26]
```diff
@@ -15,17 +15,23 @@ apiv2_auth_openid_provider:
...
 
+#add this url to maintain 4X backword compatibility
 auth_oidc_login:
-  path: /auth/oidcLogin
+  path: /openidauth/openIdCredentials
   controller: OrangeHRM\OpenidAuthentication\Controller\OpenIdConnectLoginController::handle
   methods: [ POST ]
+
+auth_oidc_login_redirect:
+  path: /openidauth/openIdCredentials/redirect
```
Comment (RajithaKumara) [inline src/plugins/orangehrmOpenidAuthenticationPlugin/config/routes.yaml:35]: I need to discuss this

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/client/src/orangehrmAuthenticationPlugin/components/SocialMediaAuth.vue (hunk 16 above)
Comment (ManulMax) [reply src/client/src/orangehrmAuthenticationPlugin/components/SocialMediaAuth.vue:23]: yes. already added in 5.1 migration

**Review by RajithaKumara — APPROVED** [review]
There are pending review feedback those should fix in the next PR

**Pull request merged**